scrapy/settings/__init__.py

```diff
@@ -88,19 +106,30 @@ def getdict(self, name, default=None):
             value = json.loads(value)
         return dict(value)
 
+    def getpriority(self, name):
+        prio = None
+        if name in self.attributes:
+            prio = self.attributes[name].priority
+        return prio
+
+    def __setitem__(self, name, value):
+        self.set(name, value)
+
     def set(self, name, value, priority='project'):
         self._assert_mutability()
-        if isinstance(priority, six.string_types):
-            priority = SETTINGS_PRIORITIES[priority]
+        priority = get_settings_priority(priority)
         if name not in self.attributes:
-            self.attributes[name] = SettingsAttribute(value, priority)
+            if isinstance(value, SettingsAttribute):
+                self.attributes[name] = value
+            else:
+                if isinstance(value, dict):
+                    value = BaseSettings(value, priority)
+                self.attributes[name] = SettingsAttribute(value, priority)
```

Comment:
After giving it some more thought, I think we should deal with BaseSettings instances explicitly instead of promoting dictionaries. That way users won't get unexpected or backward incompatible behaviors in their own dict settings, and we can choose specifically which settings are going to be BaseSettings by just declaring their values to be that.

We can still do some sort of implicit propagation only in the `Settings` class while loading the default values, checking if a setting is a dict and replacing or setting it (depending when it's done) to a BaseSettings object instead. Another option if that's not possible could it be to just import BaseSettings into `default_settings.py` and use it there for the dict values. For that matter maybe we could move `Settings` to another module within `scrapy/settings/` so we get rid of the default_settings import in scrapy/settings/\__init__.py.